lib/mayaUsd/ufe/UsdUndoMaterialCommands.cpp

```diff
@@ -0,0 +1,392 @@
+
+void UnbindMaterialUndoableCommand::undo()
+{
+    if (!_stage || _primPath.IsEmpty() || _materialPath.IsEmpty()) {
+        return;
+    }
+
+    UsdPrim          prim = _stage->GetPrimAtPath(_primPath);
+    UsdShadeMaterial material(_stage->GetPrimAtPath(_materialPath));
+    if (prim.IsValid() && material) {
+        // BindingAPI is still there since we did not remove it.
+        auto             bindingAPI = UsdShadeMaterialBindingAPI(prim);
+        UsdShadeMaterial material(_stage->GetPrimAtPath(_materialPath));
+        if (bindingAPI && material) {
+            bindingAPI.Bind(material);
+        }
+    }
+}
+
+void UnbindMaterialUndoableCommand::redo()
+{
+    if (!_stage || _primPath.IsEmpty()) {
+        return;
+    }
+
+    UsdPrim prim = _stage->GetPrimAtPath(_primPath);
+    if (prim.IsValid()) {
+        auto bindingAPI = UsdShadeMaterialBindingAPI(prim);
+        if (bindingAPI) {
+            auto materialBinding = bindingAPI.GetDirectBinding();
+            _materialPath = materialBinding.GetMaterialPath();
+            if (!_materialPath.IsEmpty()) {
+                bindingAPI.UnbindDirectBinding();
+                // TODO: Can we remove the BindingAPI at this point?
+                //       Not easy to know for sure.
+            }
+        }
+    }
+}
+const std::string UnbindMaterialUndoableCommand::commandName("Unbind Material");
+
+#if (UFE_PREVIEW_VERSION_NUM >= 4010)
+UsdUndoAssignNewMaterialCommand::UsdUndoAssignNewMaterialCommand(
+    const UsdSceneItem::Ptr& parentItem,
+    const std::string&       nodeId)
+    : Ufe::InsertChildCommand()
+    , _parentPath(parentItem->path())
+    , _nodeId(nodeId)
+{
+}
+
+UsdUndoAssignNewMaterialCommand::~UsdUndoAssignNewMaterialCommand() { }
+
+UsdUndoAssignNewMaterialCommand::Ptr UsdUndoAssignNewMaterialCommand::create(
+    const UsdSceneItem::Ptr& parentItem,
+    const std::string&       nodeId)
+{
+    // Changing the hierarchy of invalid items is not allowed.
+    if (!parentItem || !parentItem->prim().IsActive())
+        return nullptr;
+
+    return std::make_shared<UsdUndoAssignNewMaterialCommand>(parentItem, nodeId);
+}
+
+Ufe::SceneItem::Ptr UsdUndoAssignNewMaterialCommand::insertedChild() const
+{
+    return _createShaderCmd ? _createShaderCmd->insertedChild() : nullptr;
+}
+
+namespace {
+// This class insures that either everything has succeeded, or makes sure to bring back to the
+// initial state:
+class UndoGuard
+{
+public:
+    explicit UndoGuard(Ufe::UndoableCommand* cmd)
+        : _this(cmd)
+    {
+    }
+
+    ~UndoGuard()
+    {
+        // Incomplete. Undo all stored tasks until we get back to original state.
+        if (_this) {
+            _this->undo();
+        }
+    }
+    void markAsCompleted() { _this = nullptr; }
+
+private:
+    Ufe::UndoableCommand* _this = nullptr;
+};
+
+static const std::string kDefaultMaterialScopeName("mtl");
+} // namespace
+
+void UsdUndoAssignNewMaterialCommand::execute()
```

Comment:
Feels like this should be a Ufe::CompositeUndoableCommand.

Reply:
I used `has a` because `is a` leads to a diamond shape inheritance with InsertChildCommand.